Map 100 - GetFeature

wchen329 opened the pull request:
In this pull request:
* After pulling DescribeFeatureType down, GetFeature now works (third point below)
* Fixed a few typographic errors in DescribeFeatureType and changed naming of the placeholder to be consistent between GetCapabilities, DescribeFeatureType, and GetFeature.
* QGIS is able to send SQL requests (though not yet implemented) as well as add layer including the one feature type in GetFeature. Currently it is a point since points are simplest. In addition, it will draw the point in the huge expanse of whiteness.

Commit:
WFS GetFeature preliminary support

Commit:
Get Feature template implemented

Commit:
Merge

Commit:
GetFeature works!

Commit:
Made WFSGetFeatureWriter more concise.

cschroedl-gov commented on tds/src/main/java/thredds/server/wfs/WFSController.java:
```diff
@@ -185,10 +201,15 @@ public void httpHandler(HttpServletRequest hsreq, HttpServletResponse hsres) {
 					service = hsreq.getParameter(paramName);
 				}
+				
+				if(paramName.equalsIgnoreCase("TYPENAMES")) {
```
Some programmers put the literal value on the left-hand side of these kinds of comparisons to increase null safety.
https://rules.sonarsource.com/java/tag/bad-practice/RSPEC-1132
Do you like that technique?

Not a blocker for merging, just food-for-thought.

wchen329 replied:
Hmm... don't know, is null safety that big of an issue here? paramName is the identifier for the current in the for each loop, and for each is made possible by the implementation of Iterator<E>. So paramName being null is pretty much a non-possibility so here is looks like, so far is there is no bugs in the native Java libraries.

cschroedl-gov commented on tds/src/main/java/thredds/server/wfs/WFSGetFeatureWriter.java:
```diff
@@ -15,14 +15,50 @@
-				+ " xmlns:xlink=\"http://www.w3.org/1999/xlink\" xmlns:gml=\"http://opengis.net/gml\" xmlns:fes=\"http://www.opengis.net/fes/2.0\" xmlns:ogc=\"http://www.opengis.net/ogc\""
-				+ " xmlns:ows=\"http://www.opengis.net/ows/1.1\" xmlns:wfs=\"http://opengis.net/wfs/2.0\" xmlns=\"http://www.opengis.net/wfs/2.0\" version=\"2.0.0\">";
+				+ " xmlns:xlink=\"http://www.w3.org/1999/xlink\" xmlns:gml=\"http://opengis.net/gml/3.2\" xmlns:fes=\"http://www.opengis.net/fes/2.0\" xmlns:ogc=\"http://www.opengis.net/ogc\""
+				+ " xmlns:wfs=\"http://opengis.net/wfs/2.0\" xmlns:tds=\"http://localhost:8080/thredds/wfs/results/tds\" xmlns=\"http://www.opengis.net/wfs/2.0\" version=\"2.0.0\" numberMatched=\"1\" numberReturned=\"1\">"
```
This is a good start on the GetFeature response! The XML namespaces and the `xsi:schemaLocation` values might need some tweaking before they are 100% correct. It is possible some WFS clients (e.g. QGIS, OpenLayer.) will render things fine anyway without these improvements -- it depends on how rigorously the WFS clients validate the XML.

XML namespaces are kinda complicated. MSDN has a pretty good explanation here:
https://msdn.microsoft.com/en-us/library/ms256100(v=vs.110).aspx
This SO post seems relevant too:
https://stackoverflow.com/a/5875404

I'm open to suggestions on how to level up the correctness of XML schema stuff in here. One way to level up, using the GeoServer examples as inspiration, is to change the `xmlns:tds` value to a URL involving the current server's hostname (as determined from the request rather than hardcoded) and the current feature. Once that was done, a DescribeFeatureType URL could be built and then added to the value of xsi:schemaLocation.

This example looks relevant:
https://cida.usgs.gov/quality/rivers/service/geoserver/NAR/wfs?service=WFS&version=2.0.0&request=GetFeature&typeName=NAR:gtlakes_alb&count=1

```
xmlns:NAR="http://cida.usgs.gov/NAR"
```

In that case the full `xsi:schemaLocation` is:

```
xsi:schemaLocation="http://www.opengis.net/wfs/2.0 https://cida.usgs.gov:443/quality/rivers/service/geoserver/schemas/wfs/2.0/wfs.xsd http://cida.usgs.gov/NAR https://cida.usgs.gov:443/quality/rivers/service/geoserver/NAR/wfs?service=WFS&version=2.0.0&request=DescribeFeatureType&typeName=NAR%3Agtlakes_alb http://www.opengis.net/gml/3.2 https://cida.usgs.gov:443/quality/rivers/service/geoserver/schemas/gml/3.2.1/gml.xsd
```

The most relevant namespace-schema pair in that long list is the NAR namespace URL and the DescribeFeatureType URL (which returns an XML Schema):
```
http://cida.usgs.gov/NAR https://cida.usgs.gov:443/quality/rivers/service/geoserver/NAR/wfs?service=WFS&version=2.0.0&request=DescribeFeatureType&typeName=NAR%3Agtlakes_alb
```

There might be other ways to do this. I haven't looked at MapServer example WFS responses. What do you think?

wchen329 replied:
Yeah will look into it,  there's a lot of schemas and namespacing occuring in these responses!

cschroedl-gov replied:
Thanks! Yeah, they show up a lot in standards-based web services because most standards are composed of other standards which are composed of other standards... at some point the recursion stops, but it can be deeeep.

wchen329 replied:
Yeah it's a deep cavern system that can be pretty gritty to navigate!

cschroedl-gov commented on tds/src/main/java/thredds/server/wfs/WFSController.java:
```diff
@@ -23,6 +23,8 @@
 @RequestMapping("/wfs")
 public class WFSController extends HttpServlet {
 
+	public static final String TDSNAMESPACE = "\"http://localhost:8080/thredds/wfs/results\"";
```
An XML namespace should be a globally unique identifier. This is achieved by piggybacking on existing URL/DNS conventions. Since an organization controls their domain name, they can reasonably expect that no one else will specify a unique URL on their domain as an XML namespace. Since 'localhost' is a special hostname that is run by many individuals and organizations, it does not satisfy the global uniqueness constraint. Could namespace URLs in the WFS responses be built dynamically from information in each HttpServletRequest? Or based on something in the TDS catalog?

https://softwareengineering.stackexchange.com/questions/122002/why-do-we-need-uris-for-xml-namespaces

cschroedl-gov replied:
... Or could the namespace URLs be built from the `server` instance variable that many WFS*Writer classes have?

wchen329 replied:
Whoops, forgot to change that! Didn't mean to leave it as localhost, it was a "placeholder" just for the timebeing. Will update accordingly based off of HTTPServletRequest URL parameters. Thanks for catching that!

cschroedl-gov replied:
No worries! Sounds good!

cschroedl-gov reviewed (changes requested):
This is a great start! Congrats on getting this to render in QGIS! I'm most keen on getting the TDS XML namespace to be dynamic and unique. "xsi:schemaLocation" improvements can come later.

Commit:
Large refactor based off of good feedback, not 100% done but almost

Commit:
Even larger refactor

wchen329 commented:
Changes have been made.
* Fixed namespacing issues.
* Refactored other parts of Writers, namely, the construction process and included a mini XML utility class.

cschroedl-gov commented on tds/src/main/java/thredds/server/wfs/WFSController.java:
```diff
@@ -23,7 +23,27 @@
+	 * @return the namespace value as "SERVER/PATH"
+	 */
+	public static String getXMLNamespaceXMLNSValue(HttpServletRequest hsreq) {
+		return hsreq.getScheme() + "://" + hsreq.getServerName() + ":" + hsreq.getServerPort() + "/thredds/wfs/geospatial";
```
Should this delegate to `constructServerPath()`?

wchen329 replied:
Oh yeah, sure, will change it to partially delgate to the server path. This was a result of copying and pasting...!

cschroedl-gov commented on tds/src/main/java/thredds/server/wfs/WFSGetFeatureWriter.java:
```diff
@@ -48,17 +75,14 @@ public void writeMembers() {
-					+ "<gml:pos>50.0 50.0</gml:pos>"
-					+ "</gml:Point>"
+					+ "<" + WFSController.TDSNAMESPACE + ":hru_soil_moist gml:id=\"hru_soil_moist.1\">"
+					+ "<" + WFSController.TDSNAMESPACE + ":catchments_geometry_container>"
```
Good call to create and use this namespace constant!

wchen329 replied:
Thanks! Yeah allow allows for that flexibility to change in the future.

cschroedl-gov commented on tds/src/main/java/thredds/server/wfs/WFSXMLGeneral.java:
```diff
@@ -0,0 +1,26 @@
+ * @author wchen@usgs.gov
+ *
+ */
+public class WFSXMLGeneral {
```
Few classes are named "General"s. To help subsequent programmers, could you pick a more widely-understood term for the name? Would "Util" or "Utility" fit? Would "Helper"?

wchen329 replied:
Sure rename to "Helper"

cschroedl-gov commented on tds/src/main/java/thredds/server/wfs/WFSGetFeatureWriter.java:
```diff
@@ -14,16 +17,29 @@
-				+ " xmlns:xlink=\"http://www.w3.org/1999/xlink\" xmlns:gml=\"http://opengis.net/gml/3.2\" xmlns:fes=\"http://www.opengis.net/fes/2.0\" xmlns:ogc=\"http://www.opengis.net/ogc\""
-				+ " xmlns:wfs=\"http://opengis.net/wfs/2.0\" xmlns:tds=\"http://localhost:8080/thredds/wfs/results/tds\" xmlns=\"http://www.opengis.net/wfs/2.0\" version=\"2.0.0\" numberMatched=\"1\" numberReturned=\"1\">"
+	private void writeHeadersAndBB() {
+		fileOutput += "<wfs:FeatureCollection xsi:schemaLocation=" + WFSXMLGeneral.encQuotes("http://www.opengis.net/wfs/2.0 http://schemas.opengis.net/wfs/2.0/wfs.xsd " + namespace + " " + server + "?request=DescribeFeature" + WFSXMLGeneral.AMPERSAND + "service=wfs" + WFSXMLGeneral.AMPERSAND + "version=2.0.0")
```
Good use of variables and utilities in building the schemaLocation here. This schemaLocation should contain a valid `DescribeFeatureType` URL. Could you please revise the URL?

http://cite.opengeospatial.org/pub/cite/files/edu/wfs/text/operations.html#describefeaturetype

wchen329 replied:
Yeah, corrected to DescribeFeatureType and typename is set to TDSNAMESPACE:hru_soil_moist

Commit:
More changes based on good feedback

wchen329 committed:
Merge pull request #3 from wchen329/MAP-104

More changes based on good feedback
